firmware/ValdisFirmwareStuff/Core/Src/main.c

```diff
@@ -0,0 +1,771 @@
+	int16_t speed1;
+	int16_t speed2;
+	int16_t speed3;
+	uint16_t throwerSpeed;
+	uint16_t delimiter;
+} Command;
+
+typedef struct Feedback {
+	int16_t speed1;
+	int16_t speed2;
+	int16_t speed3;
+	uint16_t delimiter;
+} Feedback;
+
+
+MotorInfo M1 = {
+		.speed = 0,
+		.position = 0,
+		.integraal = 0,
+		.p_gain = 500,
+		.i_gain = 0,
+		.newspeed = 0,
+		.positionChange = 0
+};
+
+MotorInfo M2 = {
+		.speed = 0,
+		.position = 0,
+		.integraal = 0,
+		.p_gain = 500,
+		.i_gain = 0,
+		.newspeed = 0,
+		.positionChange = 0
+	};
+
+MotorInfo M3 = {
+		.speed = 0,
+		.position = 0,
+		.integraal = 0,
+		.p_gain = 500,
+		.i_gain = 0,
+		.newspeed = 0,
+		.positionChange = 0
+	};
+
+
+
+
+	Command command = {.speed1 = 0, .speed2 = 0, .speed3 = 0, .throwerSpeed = 0, .delimiter = 0};
+
+	 Feedback feedback = { // (1)
+	        .speed1 = 0,
+	        .speed2 = 0,
+	        .speed3 = 0,
+	        .delimiter = 0xAAAA
+	    };
+	volatile uint8_t isCommandReceived = 0;
+
+	int32_t clmp(int16_t val, int32_t minVal, int32_t maxVal){
+		if(val>maxVal){
+			return maxVal;
+		}
+		if(val<minVal){
+			return minVal;
+		}
+		return val;
+	};
+
+
+	int32_t PID(MotorInfo* control, int16_t position){
+		control->positionChange = position - control->position;
+		control->position = position;
+		int32_t error = control->speed -control->positionChange;
+		control->integraal += error;
+		clmp(control->integraal, -70000 / control->i_gain, 70000 / control->i_gain);
+
+		return control->p_gain * error + control->i_gain * control->integraal;
+	}
+
+	void CDC_On_Receive(uint8_t* buffer, uint32_t* length) {
+	  if (*length == sizeof(Command)) {
+	    memcpy(&command, buffer, sizeof(Command));
+	    if (command.delimiter == 0xAAAA) {
+	      isCommandReceived = 1;
+	    }
+	    if (command.delimiter == 0xBBBB) {
+	    	  		driverReset = 1;
+	    		}
+	  }
+	}
+
+	void toggle_nSleep(){
+		HAL_GPIO_WritePin(SLEEP_N_GPIO_Port,SLEEP_N_Pin,0);
+		for(int i =0;i<350;i++){
+			__asm("nop");
+		}
+		HAL_GPIO_WritePin(SLEEP_N_GPIO_Port,SLEEP_N_Pin,1);
+	}
+	void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim6) {
+			int16_t pos;
+
+			int32_t pwmValue;
+
+			pos = (int16_t)TIM1->CNT;
+			pwmValue = PID(&M1, pos);
+			HAL_GPIO_WritePin(M1_DIR_A_GPIO_Port, M1_DIR_A_Pin, (pwmValue < 0) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+			TIM2->CCR1 =(uint16_t)clmp((pwmValue < 0) ? -pwmValue : pwmValue, 0 , 65535);
+
+			pos = (int16_t)TIM3->CNT;
+			pwmValue = PID(&M2, pos);
+			HAL_GPIO_WritePin(M2_DIR_A_GPIO_Port, M2_DIR_A_Pin, (pwmValue < 0) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+			TIM2->CCR2 =(uint16_t)clmp((pwmValue < 0) ? -pwmValue : pwmValue, 0 , 65535);
+
+			pos = (int16_t)TIM4->CNT;
+			pwmValue = PID(&M3, pos);
+			HAL_GPIO_WritePin(M3_DIR_A_GPIO_Port, M3_DIR_A_Pin, (pwmValue < 0) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+			TIM2->CCR3 =(uint16_t)clmp((pwmValue < 0) ? -pwmValue : pwmValue, 0 , 65535);
+
+			toggle_nSleep();
+
+
+		  // Motor control calculations can be called from here
```

Comment:
Motor control calculations are already being called, so this comment is not useful. It's meant to be more of a TODO comment.